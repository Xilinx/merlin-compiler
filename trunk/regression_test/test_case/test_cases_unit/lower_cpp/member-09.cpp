//(C) Copyright 2016-2021 Xilinx, Inc.
//All Rights Reserved.
//
//Licensed to the Apache Software Foundation (ASF) under one
//or more contributor license agreements.  See the NOTICE file
//distributed with this work for additional information
//regarding copyright ownership.  The ASF licenses this file
//to you under the Apache License, Version 2.0 (the
//"License"); you may not use this file except in compliance
//with the License.  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing,
//software distributed under the License is distributed on an
//"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
//KIND, either express or implied.  See the License for the
//specific language governing permissions and limitations
//under the License. (edited)
// Static member function tests.

struct S {
    static unsigned f(unsigned) { return 23; }
    static void g(unsigned, S *this_) {}
    void h(unsigned) {}

    // Out-of-line definition.
    static void j();
    static int i;
    void inc() {
      this->i++;
      this->f(12);
    }
};

void S::j() {}

#pragma ACCEL kernel
int main() {
    S s;
    S::j();
    s.h(32);
    s.g(32, &s);
    S::g(32, &s);
    S *sptr;
    sptr->f(32);
    sptr->h(32);
    return S::f(32);
}
