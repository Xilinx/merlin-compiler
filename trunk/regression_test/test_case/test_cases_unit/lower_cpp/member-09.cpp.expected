// Static member function tests.

struct ::S 
{
}
;
int _ZN1S1iE;

unsigned int p_ZN1S1fEj(unsigned int )
{
  return (unsigned int )23;
}

void p_ZN1S1gEjP1S(unsigned int ,struct S *this_)
{
}

void p_ZN1S1hEj(struct S *this_,unsigned int )
{
}
void p_ZN1S1jEv();

void p_ZN1S3incEv(struct S *this_)
{
  _ZN1S1iE++;
  p_ZN1S1fEj((unsigned int )12);
}

void p_ZN1S1jEv()
{
}
static void __merlin_dummy_SgClassDeclaration_struct_S_();
static void __merlin_dummy_SgMemberFunctionDeclaration_void_j_();
#pragma ACCEL kernel

int main()
{
  struct S s;
  p_ZN1S1jEv();
  p_ZN1S1hEj(&s,32);
  p_ZN1S1gEjP1S(32,&s);
  p_ZN1S1gEjP1S(32,&s);
  struct S *sptr;
  p_ZN1S1fEj(32);
  p_ZN1S1hEj(sptr,32);
  return (p_ZN1S1fEj(32));
}
